### DISCUSSION TITLE
Extend do while syntax

### DISCUSSION THREAD

**Discussion by GiottoVerducci** [dotnet/csharplang#5874] [Language Ideas] [answered]
What is the best way to write a for loop equivalent with multiline statements?

There is a programming pattern I met often that is:

1. initialize inputs and outputs
2. perform an action
3. if the result of the action tells us to stop, return the output
4. otherwise update the outputs with the action results
5. update input for next loop
6. loop

This could be written for instance:

(First version)
```
public List<T> GetAncestors(T root)
{
	// 1. initialize inputs and outputs
	var result = new List<T>();
	var current = root;

	do
	{
		// 2. perform action (let's assume it's costly)
		T parent = current.GetParent();

		// 3. if the result of the action tells us to stop, return the output
		if (parent == null) return result;
		
		// 4. otherwise update the outputs with the action results
		result.Add(parent);
		
		// 5. update input for next loop
		current = DoSomething(parent); // this could be a costly call depending on parent that we don't want perform if not necessary
	}
	// 6. loop
	while (true);
}
```

The two inelegant parts are the ```while (true)``` condition, and the ```return``` in the middle of the loop.

There are plenty of ways of writing the same thing:

(Second version)
```
public List<T> GetAncestors(T root)
{
	var result = new List<T>();
	var current = root;

	do
	{
		T parent = current.GetParent();

		if (parent != null)
		{
			result.Add(parent);
			current = DoSomething(parent);
		}
		else
		{
			current = null;
		}
	}
	while(current != null);

	return result;
}
```

Here, we duplicate the "not null" test which is not very elegant neither.

This kind of pattern is in fact a [for-loop](https://upload.wikimedia.org/wikipedia/commons/thumb/0/06/For-loop-diagram.png/220px-For-loop-diagram.png);

(Third version)

```
public List<T> GetAncestors(T root)
{
	var result = new List<T>();

	bool TryPerformAction(T current, out T value)
	{
		value = current.GetParent();
		return value != null;
	}

	for (T current = root; TryPerformAction(current, out T value); current = DoSomething(value))
	{
		result.Add(value);
	}
	return result;
}
```
		
But this is not very elegant, and gets ugly when the initialization and the iterator are not a single statement.

Ideally, we could write something like:

``` 
public List<T> GetAncestors(T root)
{
	var result = new List<T>();

	T current = root;
	do
	{
		T parent = current.GetParent();
	}
	while (parent != null)
	{
		result.Add(parent);
		current = DoSomething(parent);
	}
	return result;
}
```

I think this would fit well in the current grammar and could be useful. What do you think?

**Comment by HaloFour** [comment]
Why do you need the extra block at all?

```cs
public List<T> GetAncestors(T root)
{
	var result = new List<T>();

	T current = root;
	T parent = current.GetParent();
	while (parent != null)
	{
		result.Add(parent);
		current = DoSomething(parent);
	}
	return result;
}
```

**Reply by canton7** [reply]
`T parent = current.GetParent();` is run on every loop

**Reply by HaloFour** [reply]
I see.  So then what's the problem with `break`?  That's why it exists.

**Reply by HaloFour** [reply]
Or this:

```cs
public List<T> GetAncestors(T root)
{
	var result = new List<T>();

	T current = root;
	T parent = current.GetParent();
	while (parent != null)
	{
		result.Add(parent);
		current = DoSomething(parent);
		parent = current.GetParent();
	}
	return result;
}
```

**Reply by HaloFour** [reply]
The suggested syntax has the second problem of declaring `parent` within the `do` scope and expecting it to be available in the `while` scope which breaks the lexical scoping rules for C#.

**Reply by GiottoVerducci** [reply]
1/ having a infinite loop with a break in it is not that great because what's the point in having an exit condition that is always true?
Imagine that for some syntaxic reason, I'd need to write ```if (true) { ... }``` to achieve something, I don't feel it's elegant.
Here, I have to write ```while (true) { ... }``` to achieve something.

2/ In your example, the code that sets parent is duplicated. It's not really elegant neither.

3/ I understand there's a scope issue, that's why it's hard to implement it with an utility method (you'd have to pass some parameters back and forth)
The lexical scoping is already lose for some cases such as variable declarations into a statement (e.g., 
```for(int i = 0;.. { ... use i here }```. 
or 
```
if(TryGetValue(out int value)) 
{
    // .... use value here 
}
// or even outside the scope
if(value == 3)
{
}
```
I don't think this would be a big deal.

**Reply by CyrusNajmabadi** [reply]
> Here, I have to write while (true) { ... } to achieve something.

If you don't like `while (true)` you could always just `goto some_label;` to achieve the same.

**Comment by ufcpp** [comment]
```cs
	while (true)
	{
		T parent = current.GetParent();

		if (parent is null) break;

		result.Add(parent);
		current = DoSomething(parent);
	}
```

**Reply by canton7** [reply]
Surely this is just a copy/paste of OP's "First version" code?

**Reply by GiottoVerducci** [reply]
Yes, the ```while(true)... break;``` is just a variant of First Version. You still have a infinite loop and a break, which works but is not very elegant.

**Comment by TahirAhmadov** [comment, marked as answer]
The `while(true) { ... if(...) break; ... }` workaround is essentially a tiny-bit-better looking `goto` setup. I would also a "middle-condition loop" (as opposed to `while` which is "start condition" and `do` which is "end condition"). Maybe:
```cs
do
{
  ...
  continue if (...);
  ...
}
```

**Reply by bernd5** [reply]
Does it really make anything better compared to `if (condition) break;`?

**Reply by TahirAhmadov** [reply]
I think it's just readability and well-defined structure we're hoping to get here. Yes, `while(true)` and `if(condition) break;` works, but so does a couple of `goto` with labels. I'm not stuck on what I wrote - just throwing out ideas. At some point in the past, I came up with a new keyword - `midwhile`:
```cs
do
{
   ...
   midwhile(condition);
   ...
}
```

**Reply by GiottoVerducci** [reply]
I was thinking of something similar, simply a "while" in the middle of the clause.
EDIT: you're right, we cannot reuse ```while``` since ```while(condition);``` has its own meaning.

**Comment by CyrusNajmabadi** [comment]
> The two inelegant parts are the while (true) condition, and the return in the middle of the loop.

I'm genuinely not sure what is elegant about either of those.  THe loop itself is unconditional (which makes sense to me as `while (true)`.  And something in teh loop returns out.  Both of these seem normal and reasonable.  

It's also very readable to me.  So i'm not really sure if there's an actual problem here needing a solution.

**Comment by CyrusNajmabadi** [comment]
Feels like you can just write this as:

```c#
var result = new List<T>();
for (var current = root.GetParent(); current != null; current = DoSomething(current).GetParent())
{
    result.Add(current);
}

return result;
```

**Reply by GiottoVerducci** [reply]
See "Third version", it gets ugly when the initialization and the iterator are not single statements.

**Reply by CyrusNajmabadi** [reply]
Doesn't appear to be ugly to me.

**Reply by FaustVX** [reply]
Honestly, the Third Version looks way cleaner and readable than mid-loop `while`.

**Comment by Richiban** [comment]
Don't forget you can use pattern matching in the condition section of the `while` loop:

```cs
public List<T> GetAncestors(T root)
{
    var result = new List<T>();
    var current = root;

    while (current.GetParent() is {} parent)
    {
        result.Add(parent);
        current = DoSomething(parent);
    }

    return result;
}

```

I do this all this time. It also works nicely with things like queues, where you can do:

```cs
Queue<T> queue = ...

while (queue.TryDequeue(out var current))
{
    // Now I have access to the item that was just dequeued
}
```
